### DISCUSSION TITLE
Support #autoawait directive and corresponding 'noawait' operator

### DISCUSSION THREAD

**Discussion by craigajohnson** [dotnet/csharplang#6060] [Language Ideas]
With the asyncify-everything approach pervasive, a lot of unnecessary syntactic noise is generated.

The rules for #autoawait would be simple:

When `#autoawait enable` is specified (whether within a .cs file or by csproj property default):
- An expression that has a valid GetAwaiter() would be automatically awaited
- If you want the awaitable thing instead, specify `noawait`

Example:

```
#autoawait enable

Task.Delay(100); // compiler lowers to await automatically

var x = noawait Task.Delay(100); // assigns Task to x

await Task.Delay(100) // analyzer issues warning '#autoawait is enabled; await operator is not necessary'
```

While this would clean up noise generally, this would have particular value in projects such as UI tests using async test frameworks, where nearly every line of test methods begin with `await`.

Use of the new directive would be opt-in, similar to other recent directives (e.g., #nullable), preserving backcompat.

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/5821
https://github.com/dotnet/csharplang/discussions/5469

**Reply by craigajohnson** [reply]
@HaloFour I follow and appreciate your work but I'm going to respectfully disagree as I'm not advocating for the same thing here:

**`#autoawait` is a call-site change only. This is not Go-ifying the stack. Use of the `async` modifier would necessarily continue**

Sufficient delineation between async (red) and sync (blue) is a genuine concern and I can see how this proposal could blur those boundaries. As with all things, judiciousness is key. For certain types of code, the readability advantage of `#autoawait` would outweigh the disadvantage of over-hiding complexity.

If the concern is that enabling blanket-level `#autoawait` would muddy the waters, perhaps an _await block_ could solve that?

(Example UI automation script):

```
await
{
   var h = GetHomePageAsync();

   h.Account.SetTextAsync("abc");
   h.Password.SetTextAsync("***");
   h.LoginButton.ClickAsync();

   // 200 more lines no longer requiring individual await operators...
}
```

The goal is not to conflate sync vs. async, but rather to eliminate unnecessary noise in certain targeted scenarios.

**The `noawait` operator declares the developer's intent and understanding of what's going on when returning a Task/ValueTask/other awaitable is desired**

In a sense, `#autoawait` just inverts the default behavior for dealing with expressions having a suitable `GetAwaiter()`. In those situations where let's say it's 90% await Task, 10% return a Task, use of `noawait` makes it clear that "hey I know what's happening here - I've chosen to reduce clutter for the 90%, but I'm adding this 10% to a List and I'll `noawait` those for my own wise purposes"

There will always be a competing tension between the productivity of making hard things magically happen (e.g., allocations/GC, async, iterators, etc.) and the reality of what is actually going on under the hood so engineers can make reasoned decisions. Helping the community to understand these trade-offs I'm sure is a key design consideration and an enormous challenge.

**Reply by HaloFour** [reply]
You're advocating for making `await` implicit, which is contradictory to the design goals of `async`/`await`.  The team members have been pretty clear that they want the coroutine boundaries to be explicit and apparent.  That boundary represents I/O and thread switching which could have substantial impact on the behavior of that code.  An option to reverse that design not only hides the boundary, but it also introduces a dialect within the language.

**Reply by craigajohnson** [reply]
The await block in the example above hides nothing - it simply reduces noise while maintaining explicit intent.

**Reply by CyrusNajmabadi** [reply]
Note: this makes it a breaking change to add an extension GetAwaiter in a namespace. Now the consumption side that has imported that namespace will change meaning on any code that isn't using `await` on a value that has now become awaitable.

**Reply by CyrusNajmabadi** [reply]
I see the desire for this feature. But I think there around be problems. For example, how to ConfigureAwait all the implicit await calls. 

As halo mentioned above, there is a lot of value in our minds in seeing the awaiting being explicit at the point it happens. I worry about lack of clarity if this becomes a block construct and those switches are no longer clear.

**Reply by craigajohnson** [reply]
@CyrusNajmabadi @HaloFour thank you for constructive comments as always.

> Note: this makes it a breaking change to add an extension GetAwaiter in a namespace. Now the consumption side that has imported that namespace will change meaning on any code that isn't using `await` on a value that has now become awaitable.

I agree, although it shouldn't be breaking per se as usage is opt-in, especially for the `await` block syntax which is more explicit

> I see the desire for this feature. But I think there around be problems. For example, how to ConfigureAwait all the implicit await calls.
> 
> As halo mentioned above, there is a lot of value in our minds in seeing the awaiting being explicit at the point it happens. I worry about lack of clarity if this becomes a block construct and those switches are no longer clear.

Understood. I wonder though if in some ways, the `await` block could effectively solve some languishing issues around default behavior of ConfigureAwait. In some ways it could be considered a scoping tool to specify CA preference.

```
await  // ConfigureAwait(true) would still be the default - all calls are ConfigureAwait(true) and maintain context
{
   var h = GetHomePageAsync(); // UI automation script should be on same thread, so default works best here anyways
   h.UserName.SetTextAsync("grover");
   h.LoginAsync();

   // 200 more lines of UI script...
}
```

or

```
await(false)
{
   var sr = Service.LoadAsync();    // lowered to await Service.LoadAsync().ConfigureAwait(false);
   var xr = ReturnToThisContextAsync().ConfigureAwait(true);   // compiler honors explicit choice

   // more
}
```

I see usage being for targeted scenarios to reduce noise while maintaining explicit intent.

**Comment by TahirAhmadov** [comment]
Please look at my proposal at https://github.com/dotnet/csharplang/discussions/5782. It's a different take but I think it's pursuing a similar goal.

**Reply by craigajohnson** [reply]
I wouldn't be able to support that I'm afraid as it is a massively intrusive proposal with several incomplete assumptions. I could not support automagically changing method names behind the scenes and in fact prefer to have the Async suffix and agree with the threading analyzer that generates a warning if a method with an async modifier is not suffixed with Async.

I would have a look at the `await` operator in the C# language specification to get an idea of its usage and applicable scope and its direct association with `GetAwaiter()`
